blog/2025/keycloak-bencmark.adoc

```diff
@@ -0,0 +1,198 @@
+
+== Scaling Keycloak
+
+One of the most common questions when deploying Keycloak is: **how many resources do I need?**
+The answer, as you may have guessed, is that it depends on your specific use case.
+
+For this test, we scaled only the login and refresh token requests.
+Based on Keycloak's https://www.keycloak.org/high-availability/concepts-memory-and-cpu-sizing[Concepts for sizing CPU and memory resources] documentation, we determined that you need **1 vCPU to handle 15 logins per second** and an additional **1 vCPU to handle 120 refresh token requests per second**.
+
+After computing the total number of vCPUs required, we divided the value by the desired number of Pods, which was three in this case (one for each availability zone).
+We allocated slightly more vCPU to each Pod than the calculated value to account for JVM background tasks like compilations and garbage collection.
+
+[IMPORTANT]
+====
+Because our test used a constant load, we did not allocate as much CPU as we typically recommend.
+We **strongly recommend** leaving **150% extra headroom** for CPU usage to handle spikes in load, as mentioned in our documentation.
+====
+
+Memory was increased when we observed high CPU usage by the garbage collector (GC).
+It is difficult to formulate a precise memory requirement because it depends not only on the Keycloak data but also on the number of concurrent requests.
+
+The database instance type was chosen based on observation.
+If we observed failing requests or a 99th percentile response time above 100ms with database CPU usage exceeding 80%, we repeated the test using the next larger database instance type.
+
+A summary of the results can be observed in the table below.
+
+.Keycloak performance with load
+|===
+|OCP Instance Type | DB Instance Type | Pod CPU limit | Pod Memory Limit (GB) | # Pods | Logins/sec | Refreshes Token/sec| Response Time (ms, 99th pct)
+
+m|c8g.8xlarge
+m|db.r8g.2xlarge
+|24
+|4
+|3
+|500
+|2500
+|81
+
+m|c8g.8xlarge
+m|db.r8g.4xlarge
+|40
+|8
+|3
+|1000
+|5000
+|60
+
+m|c8g.24xlarge
+m|db.r8g.16xlarge
+|74
+|8
+|3
+|2000
+|10000
+|58
+
+m|c8g.24xlarge
+m|db.r8g.16xlarge
+|40
+|8
+|6
+|1000
+|20000
+|59
+
+|===
+
+Based on the overall results, we can confirm that Keycloak scales vertically almost linearly.
+
+We used the default Keycloak settings, with the exception of the last two tests.
+For those tests, we had to increase the number of threads (using the `http-pool-max-threads` option) to 330 for the 2,000 logins + 10,000 token refreshes scenario and to 200 for the 1,000 logins + 20,000 token refreshes scenario.
+
+It is also worth noting that we scaled up the OpenShift Ingress Routers because we were observing connection errors on the load generator.
+
+The following images illustrate how the requests correlated with the configured CPU and memory limits.
+
+.CPU limits
+image:${blogImages}/kc_perf_2025/kc_perf_1.png[requests vs cpu]
```

Comment:
The referenced graph has a kink. But it is due to the change from logins to refreshes, not due to a scaling. So I think it is misleading, and would consider removing it.

Reply:
Sure, I'll remove it.